File: code/com/leetcode/solutions/NO_0006_Z字形变换/Solution.cs
using System;

public class Solution
{
    /*
  * s:指定字符串
  * numRows:指定行数
  * 思路：每一行的字符用数组存储，遍历字符串，通过变量判断字符应该存储到哪一个数组中，注意问题。
 */
    public string Convert(string s, int numRows)
    {
        StringBuilder res = new StringBuilder();
        if (numRows == 1)
        {

            return s;
        }
        else
        {

            Dictionary<int, List<char>> allData = new Dictionary<int, List<char>>();
            for (int i = 0; i < numRows; i++)
            {
                allData.Add(i, new List<char>());
            }
            int strIndex = 0;
            bool gor = true;
            for (int j = 0; j < s.Length; j++)
            {
                allData[strIndex].Add(s[j]);
                if (gor)
                {
                    strIndex++;
                }
                else
                {
                    strIndex--;
                }

                if (strIndex == 0)
                {
                    gor = true;
                }
                else if (strIndex == numRows - 1)
                {
                    gor = false;
                }
            }

            for (int k = 0; k < allData.Count; k++)
            {
                string item = string.Join("", allData[k]);
                res.Append(item);
            }
        }

        return res.ToString();

    }
}

File: code/com/leetcode/solutions/NO_0006_Z字形变换/Solution.java
package com.leetcode.solutions.NO_0006_Z字形变换;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by gang.xu01@hand-china.com on 2018/12/27
 */
public class Solution {

    /**
     * 解法：行的思想，按照行把所有值塞到list里面，在一次遍历数组的时候，标注遍历方向，按照方向和顺序塞入到各自的list当中
     *      最后输出list即可
     *
     * @param s 输入字符串
     * @param numRows 显示多少行
     * @return
     */
    public String convert(String s, int numRows) {

        if (numRows <= 0) throw new RuntimeException("error");
        else if (numRows == 1) return s;

        StringBuilder returnStr = new StringBuilder();
        List<StringBuffer> sbList = new ArrayList<>();
        for (int i = 0; i < Math.min(numRows, s.length()); i++) {
            sbList.add(new StringBuffer());
        }

        boolean down = false;
        int row = 0;
        for (char c : s.toCharArray()) {
            sbList.get(row).append(c);
            if(row == 0 || row == numRows - 1) down = !down;
            if(down) row++; else row--;
        }

        for (StringBuffer sb : sbList) {
            returnStr.append(sb);
        }

        return returnStr.toString();
    }


    public static void main(String[] args) {

    }

}
